runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkProcessContext.java

```diff
@@ -61,17 +62,13 @@
   Iterable<OutputT> processPartition(
       Iterator<WindowedValue<FnInputT>> partition) throws Exception {
 
-    // setup DoFn.
-    DoFnInvokers.invokerFor(doFn).invokeSetup();
-
     // skip if partition is empty.
     if (!partition.hasNext()) {
-      DoFnInvokers.invokerFor(doFn).invokeTeardown();
       return Lists.newArrayList();
     }
 
-    // call startBundle() before beginning to process the partition.
```

Comment:
Why not just call setup here, and wrap (try/catch) `startBundle` call and `getOutputIterable`, and in the catch clause call teardown and re-throw?
Looks simpler, unless I'm missing something.

Reply:
Ok, I will check this and try to get into a readability compromise.